apps/rush-lib/src/api/CommandLineConfiguration.ts

```diff
@@ -29,270 +96,399 @@ export class CommandLineConfiguration {
-            if (!this.phases.has(dependency)) {
-              throw new Error(
-                `In ${RushConstants.commandLineFilename}, in the phase "${phase.name}", the upstream ` +
-                  `dependency phase "${dependency}" does not exist.`
-              );
-            }
-          }
-        }
+      this._checkForPhaseSelfCycles(phase);
+      const phaseClique: Set<string> = new Set<string>();
+      this._populatePhaseClique(phase.name, phaseClique);
+      phaseCliques.set(phase.name, phaseClique);
+    }
 
-        this._checkForPhaseSelfCycles(phase);
+    function populateCommandsForPhase(phaseName: string, command: IPhasedCommand): void {
+      const phaseClique: Set<string> = phaseCliques.get(phaseName)!;
+      for (const phaseCliqueIdentifier of phaseClique) {
+        commandsForPhase.get(phaseCliqueIdentifier)!.add(command);
       }
+    }
 
-      if (commandLineJson.commands) {
-        for (const command of commandLineJson.commands) {
-          if (this.commands.has(command.name)) {
-            throw new Error(
-              `In ${RushConstants.commandLineFilename}, the command "${command.name}" is specified ` +
-                'more than once.'
-            );
-          }
+    // A map of bulk command names to their corresponding synthetic phase identifiers
+    const syntheticPhasesForTranslatedBulkCommands: Map<string, string> = new Map<string, string>();
+    const translateBulkCommandToPhasedCommand: (
+      command: IBulkCommandJson,
+      isBuildCommand: boolean
+    ) => IPhasedCommand = (command: IBulkCommandJson, isBuildCommand: boolean) => {
+      const phaseName: string = command.name;
+      const phaseForBulkCommand: IPhase = {
+        name: phaseName,
+        isSynthetic: true,
+        dependencies: {
+          upstream: command.ignoreDependencyOrder ? undefined : [phaseName]
+        },
+        ignoreMissingScript: command.ignoreMissingScript,
+        allowWarningsOnSuccess: command.allowWarningsInSuccessfulBuild,
+        logFilenameIdentifier: this._normalizeNameForLogFilenameIdentifiers(command.name),
+        // Because this is a synthetic phase, just use the project name because there aren't any other phases
+        getDisplayNameForProject: (rushProject: RushConfigurationProject) => rushProject.packageName
+      };
+      this.phases.set(phaseName, phaseForBulkCommand);
+      syntheticPhasesForTranslatedBulkCommands.set(command.name, phaseName);
+      const phaseClique: Set<string> = new Set<string>();
+      this._populatePhaseClique(phaseName, phaseClique);
+      phaseCliques.set(phaseName, phaseClique);
+
+      const translatedCommand: IPhasedCommand = {
+        ...command,
+        commandKind: 'phased',
+        disableBuildCache: true,
+        isSynthetic: true,
+        associatedParameters: new Set<Parameter>(),
+        phases: [phaseName]
+      };
+      commandsForPhase.set(phaseName, new Set<IPhasedCommand>());
```

Comment:
What if the user tries to define two phases that have the same phase name? Do we validate that somewhere?

Reply:
We throw

```TypeScript
        if (this.phases.has(phase.name)) {
          throw new Error(
            `In ${RushConstants.commandLineFilename}, the phase "${phase.name}" is specified ` +
              'more than once.'
          );
        }
```